target_chains/ethereum/contracts/contracts/wormhole-receiver/ReceiverMessages.sol

```diff
@@ -7,20 +7,173 @@ pragma experimental ABIEncoderV2;
+                encodedVM,
+                index
+            );
+            index += 1;
+            {
+                uint hashIndex = index + (signersLen * 66);
+                if (hashIndex > encodedVM.length) {
+                    return (vm, false, "invalid signature length");
+                }
+                // Hash the body
+                vm.hash = keccak256(
+                    abi.encodePacked(
+                        keccak256(
+                            UnsafeCalldataBytesLib.sliceFrom(
+                                encodedVM,
+                                hashIndex
+                            )
+                        )
+                    )
+                );
+            }
+
+            {
+                uint8 lastIndex = 0;
+                for (uint i = 0; i < signersLen; i++) {
+                    ReceiverStructs.Signature memory sig;
+                    sig.guardianIndex = UnsafeCalldataBytesLib.toUint8(
+                        encodedVM,
+                        index
+                    );
+                    index += 1;
+
+                    sig.r = UnsafeCalldataBytesLib.toBytes32(encodedVM, index);
+                    index += 32;
+                    sig.s = UnsafeCalldataBytesLib.toBytes32(encodedVM, index);
+                    index += 32;
+                    sig.v =
+                        UnsafeCalldataBytesLib.toUint8(encodedVM, index) +
+                        27;
+                    index += 1;
+
+                    bool signatureValid;
+                    string memory invalidReason;
+                    (signatureValid, invalidReason) = verifySignature(
+                        i,
+                        lastIndex,
+                        vm.hash,
+                        sig,
+                        guardianSet
+                    );
+                    if (!signatureValid) {
+                        return (vm, false, invalidReason);
+                    }
+                    lastIndex = sig.guardianIndex;
+                }
+            }
+
+            /**
+             * @dev We're using a fixed point number transformation with 1 decimal to deal with rounding.
+             *   WARNING: This quorum check is critical to assessing whether we have enough Guardian signatures to validate a VM
+             *   if making any changes to this, obtain additional peer review. If guardianSet key length is 0 and
+             *   vm.signatures length is 0, this could compromise the integrity of both vm and signature verification.
+             */
+
+            if (
+                (((guardianSet.keys.length * 10) / 3) * 2) / 10 + 1 > signersLen
+            ) {
+                return (vm, false, "no quorum");
+            }
+
+            // purposely setting vm.signatures to empty array since we don't need it anymore
+            // and we've already verified it above
+            vm.signatures = new ReceiverStructs.Signature[](0);
```

Comment:
this scares me a little because the overall function is no longer a drop-in replacement for the old one. If anyone was depending on the old behavior, then this could break them.

How much of a gas savings was this change? If it wasn't big, may be worthwhile to undo this.

Reply:
The gas saving is very big. Also, this is Pyth's version of Wormhole and we know that we don't rely on it, so your concern doesn't make much sense.

Reply:
here are the numbers if we actually fill the `vm.signatures`
overall we lose about half of the gas savings

![image](https://github.com/pyth-network/pyth-crosschain/assets/86628128/97471a62-7193-47f4-b54b-2dbcaff3906c)

Reply:
i updated the code comment on `parseAndVerifyVM` to include a warning about the `vm.signatures` and to use `parseVM` and `verifyVM` if the caller wants to get a populated `vm.signatures`